drivers/dma/dma_stm32u5.c

```diff
@@ -0,0 +1,832 @@
+		    &dma_funcs)
+
+#define DMA_STM32_DEFINE_IRQ_HANDLER(dma, chan)				\
+static void dma_stm32_irq_##dma##_##chan(const struct device *dev)	\
+{									\
+	dma_stm32_irq_handler(dev, chan);				\
+}
+
+
+#define DMA_STM32_IRQ_CONNECT(dma, chan)				\
+	do {								\
+		IRQ_CONNECT(DT_INST_IRQ_BY_IDX(dma, chan, irq),		\
+			    DT_INST_IRQ_BY_IDX(dma, chan, priority),	\
+			    dma_stm32_irq_##dma##_##chan,		\
+			    DEVICE_DT_INST_GET(dma), 0);		\
+		irq_enable(DT_INST_IRQ_BY_IDX(dma, chan, irq));		\
+	} while (0)
+
+#if DT_NODE_HAS_STATUS(DT_DRV_INST(0), okay)
+
+DMA_STM32_DEFINE_IRQ_HANDLER(0, 0);
+DMA_STM32_DEFINE_IRQ_HANDLER(0, 1);
+DMA_STM32_DEFINE_IRQ_HANDLER(0, 2);
+DMA_STM32_DEFINE_IRQ_HANDLER(0, 3);
+DMA_STM32_DEFINE_IRQ_HANDLER(0, 4);
+#if DT_INST_IRQ_HAS_IDX(0, 5)
+DMA_STM32_DEFINE_IRQ_HANDLER(0, 5);
+#if DT_INST_IRQ_HAS_IDX(0, 6)
+DMA_STM32_DEFINE_IRQ_HANDLER(0, 6);
+#if DT_INST_IRQ_HAS_IDX(0, 7)
+DMA_STM32_DEFINE_IRQ_HANDLER(0, 7);
+#endif /* DT_INST_IRQ_HAS_IDX(0, 5) */
+#endif /* DT_INST_IRQ_HAS_IDX(0, 6) */
+#endif /* DT_INST_IRQ_HAS_IDX(0, 7) */
+
+DMA_STM32_DEFINE_IRQ_HANDLER(0, 8);
+DMA_STM32_DEFINE_IRQ_HANDLER(0, 9);
+DMA_STM32_DEFINE_IRQ_HANDLER(0, 10);
+DMA_STM32_DEFINE_IRQ_HANDLER(0, 11);
+DMA_STM32_DEFINE_IRQ_HANDLER(0, 12);
+DMA_STM32_DEFINE_IRQ_HANDLER(0, 13);
+DMA_STM32_DEFINE_IRQ_HANDLER(0, 14);
+DMA_STM32_DEFINE_IRQ_HANDLER(0, 15);
+
+static void dma_stm32_config_irq_0(const struct device *dev)
+{
+	ARG_UNUSED(dev);
+
+	DMA_STM32_IRQ_CONNECT(0, 0);
+	DMA_STM32_IRQ_CONNECT(0, 1);
+	DMA_STM32_IRQ_CONNECT(0, 2);
+	DMA_STM32_IRQ_CONNECT(0, 3);
+	DMA_STM32_IRQ_CONNECT(0, 4);
+#if DT_INST_IRQ_HAS_IDX(0, 5)
+	DMA_STM32_IRQ_CONNECT(0, 5);
+#if DT_INST_IRQ_HAS_IDX(0, 6)
+	DMA_STM32_IRQ_CONNECT(0, 6);
+#if DT_INST_IRQ_HAS_IDX(0, 7)
+	DMA_STM32_IRQ_CONNECT(0, 7);
+#endif /* DT_INST_IRQ_HAS_IDX(0, 5) */
+#endif /* DT_INST_IRQ_HAS_IDX(0, 6) */
+#endif /* DT_INST_IRQ_HAS_IDX(0, 7) */
+/* Either 5 or 6 or 7 or 8 channels for DMA across all stm32 series. */
+	DMA_STM32_IRQ_CONNECT(0, 8);
+	DMA_STM32_IRQ_CONNECT(0, 9);
+	DMA_STM32_IRQ_CONNECT(0, 10);
+	DMA_STM32_IRQ_CONNECT(0, 11);
+	DMA_STM32_IRQ_CONNECT(0, 12);
+	DMA_STM32_IRQ_CONNECT(0, 13);
+	DMA_STM32_IRQ_CONNECT(0, 14);
+	DMA_STM32_IRQ_CONNECT(0, 15);
+}
+
+DMA_STM32_INIT_DEV(0);
+
+#endif /* DT_NODE_HAS_STATUS(DT_DRV_INST(0), okay) */
+
+
+#if DT_NODE_HAS_STATUS(DT_DRV_INST(1), okay)
+
+DMA_STM32_DEFINE_IRQ_HANDLER(1, 0);
+DMA_STM32_DEFINE_IRQ_HANDLER(1, 1);
+DMA_STM32_DEFINE_IRQ_HANDLER(1, 2);
+DMA_STM32_DEFINE_IRQ_HANDLER(1, 3);
+#if DT_INST_IRQ_HAS_IDX(1, 4)
+DMA_STM32_DEFINE_IRQ_HANDLER(1, 4);
+#if DT_INST_IRQ_HAS_IDX(1, 5)
+DMA_STM32_DEFINE_IRQ_HANDLER(1, 5);
+#if DT_INST_IRQ_HAS_IDX(1, 6)
+DMA_STM32_DEFINE_IRQ_HANDLER(1, 6);
+#if DT_INST_IRQ_HAS_IDX(1, 7)
+DMA_STM32_DEFINE_IRQ_HANDLER(1, 7);
+#endif /* DT_INST_IRQ_HAS_IDX(1, 4) */
+#endif /* DT_INST_IRQ_HAS_IDX(1, 5) */
+#endif /* DT_INST_IRQ_HAS_IDX(1, 6) */
+#endif /* DT_INST_IRQ_HAS_IDX(1, 7) */
```

Comment:
Can it be simplified?

Reply:
yes, as long as stm32u5 or compatible soc series have one GPDMA of 15 channels